src/hmac_key.c

```diff
@@ -0,0 +1,875 @@
+	0x1f, 0x83, 0xd9, 0xab, 0xfb, 0x41, 0xbd, 0x6b, 0x5b, 0xe0, 0xcd, 0x19,
+	0x13, 0x7e, 0x21, 0x79,
+};
+
+#define SHA224_DIGEST_LENGTH          28
+#define SHA256_DIGEST_LENGTH          32
+#define SHA384_DIGEST_LENGTH          48
+#define SHA512_DIGEST_LENGTH          64
+
+int zpc_hmac_key_alloc(struct zpc_hmac_key **hmac_key)
+{
+	pthread_mutexattr_t attr;
+	struct zpc_hmac_key *new_hmac_key = NULL;
+	int rc, rv, attr_init = 0;
+
+	UNUSED(rv);
+
+	if (pkeyfd < 0) {
+		rc = ZPC_ERROR_DEVPKEY;
+		goto ret;
+	}
+	if (hmac_key == NULL) {
+		rc = ZPC_ERROR_ARG1NULL;
+		goto ret;
+	}
+
+	new_hmac_key = calloc(1, sizeof(*new_hmac_key));
+	if (new_hmac_key == NULL) {
+		rc = ZPC_ERROR_MALLOC;
+		goto ret;
+	}
+
+	rc = pthread_mutexattr_init(&attr);
+	if (rc) {
+		rc = ZPC_ERROR_MALLOC;
+		goto ret;
+	}
+	attr_init = 1;
+	rv = pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_RECURSIVE);
+	assert(rv == 0);
+	rc = pthread_mutex_init(&new_hmac_key->lock, &attr);
+	if (rc) {
+		rc = ZPC_ERROR_INITLOCK;
+		goto ret;
+	}
+	new_hmac_key->refcount = 1;
+	DEBUG("hmac key at %p: refcount %llu", new_hmac_key, new_hmac_key->refcount);
+
+	*hmac_key = new_hmac_key;
+	rc = 0;
+ret:
+	if (attr_init == 1) {
+		rv = pthread_mutexattr_destroy(&attr);
+		assert(rv == 0);
+	}
+	if (rc)
+		free(new_hmac_key);
+	DEBUG("return %d (%s)", rc, zpc_error_string(rc));
+	return rc;
+}
+
+int zpc_hmac_key_set_type(struct zpc_hmac_key *hmac_key, int type)
+{
+	int rc, rv;
+
+	UNUSED(rv);
+
+	if (pkeyfd < 0) {
+		rc = ZPC_ERROR_DEVPKEY;
+		DEBUG("return %d (%s)", rc, zpc_error_string(rc));
+		return rc;
+	}
+	if (hmac_key == NULL) {
+		rc = ZPC_ERROR_ARG1NULL;
+		DEBUG("return %d (%s)", rc, zpc_error_string(rc));
+		return rc;
+	}
+
+	switch (type) {
+	case ZPC_HMAC_KEY_TYPE_PVSECRET:
+		if (!swcaps.uv_pvsecrets) {
+			rc = ZPC_ERROR_UV_PVSECRETS_NOT_AVAILABLE;
+			DEBUG("return %d (%s)", rc, zpc_error_string(rc));
+			return rc;
+		}
+		break;
+	default:
+		rc = ZPC_ERROR_KEYTYPE;
+		DEBUG("return %d (%s)", rc, zpc_error_string(rc));
+		return rc;
+	}
+
+	rv = pthread_mutex_lock(&hmac_key->lock);
+	assert(rv == 0);
+
+	if (hmac_key->refcount != 1) {
+		rc = ZPC_ERROR_OBJINUSE;
+		goto ret;
+	}
+
+	DEBUG("hmac key at %p: type set to %d", hmac_key, type);
+	hmac_key->type = type;
+	hmac_key->type_set = 1;
+	rc = 0;
+ret:
+	rv = pthread_mutex_unlock(&hmac_key->lock);
+	assert(rv == 0);
+	DEBUG("return %d (%s)", rc, zpc_error_string(rc));
+	return rc;
+}
+
+int zpc_hmac_key_set_hash_function(struct zpc_hmac_key *hmac_key, zpc_hmac_hashfunc_t hfunc)
+{
+	int rc, rv;
+
+	UNUSED(rv);
+
+	if (pkeyfd < 0) {
+		rc = ZPC_ERROR_DEVPKEY;
+		DEBUG("return %d (%s)", rc, zpc_error_string(rc));
+		return rc;
+	}
+	if (hmac_key == NULL) {
+		rc = ZPC_ERROR_ARG1NULL;
+		DEBUG("return %d (%s)", rc, zpc_error_string(rc));
+		return rc;
+	}
+
+	switch (hfunc) {
+	case ZPC_HMAC_HASHFUNC_SHA_224:
+	case ZPC_HMAC_HASHFUNC_SHA_256:
+	case ZPC_HMAC_HASHFUNC_SHA_384:
+	case ZPC_HMAC_HASHFUNC_SHA_512:
+		break;
```

Comment:
Is it checked anywhere that the hash function set here matches the retrievable secret type (ZPC_HMAC_SECRET_HMAC_SHA_xxx) ?

Reply:
As commented above, the retrievable secret type is derived from the hash function. The import function checks if the expected secret is available in the system. For this check, the expected secret type is set in the ioctl, see routine hmac_key_make_uvrsecrettoken() in hmac_key.c.

Reply:
Ok, so if the secret ID as such is valid and exists, but the type does not match the expended type, then it fails.
Does it produce different error codes if
1.) the secret Id die sot exist at all
2.) the secret Id exists, but is a wrong type ? 
This would be important to differentiate to give the user a hit about whats wrong. 
(did not look into the code, sorry)

Reply:
Hmm, good point. At the moment the PKEY_VERIFYKEY2 ioctl just fails if the "expected" secret is not available. But I'm not checking if the given secret ID probably points to an existing secret with another type. Maybe I could change the check so that both types are tried. Then we would see the exact error.

Reply:
As discussed offline I added another commit to make the rc ZPC_ERROR_PVSECRET_ID_NOT_FOUND_IN_UV more informative.